### PR TITLE
[SPARK-20736][Python] PySpark StringIndexer supports StringOrderType

### PR DESCRIPTION
**Pull request by actuaryzhang**
## What changes were proposed in this pull request?
PySpark StringIndexer supports StringOrderType added in #17879.

### CODE REVIEW

**Comment by actuaryzhang** [comment]
@felixcheung

**Comment by actuaryzhang** [comment]
@viirya @MLnick @BryanCutler @yinxusen @brkyvz @HyukjinKwon @srowen 
Ping for reviews or comments. Thanks much.

File: python/pyspark/ml/feature.py
Comment (HyukjinKwon) [inline python/pyspark/ml/feature.py]: Probably, the leading `\` could be removed.

Comment (HyukjinKwon) [inline python/pyspark/ml/feature.py]: (Probably, the leading `\` could be removed.)

**Comment by HyukjinKwon** [comment]
(I am not used to ML. I just left a trivial comment for Python.)

**Comment by viirya** [comment]
Code changes looks good. But we need to add test for this.

File: python/pyspark/ml/feature.py
Comment (viirya) [inline python/pyspark/ml/feature.py]: alphabetDsec -> alphabetDesc

**Comment by actuaryzhang** [comment]
@viirya Thanks much for your review. I corrected the typo and added some tests.

File: python/pyspark/ml/feature.py
Comment (actuaryzhang) [inline python/pyspark/ml/feature.py]: @HyukjinKwon  Thank you. Added tests.

**Comment by actuaryzhang** [comment]
@felixcheung Could you take a look? Thanks.

File: python/pyspark/ml/feature.py
Comment (felixcheung) [inline python/pyspark/ml/feature.py]: I think this should be generated instead of hardcoded - you can find a example on python..

File: python/pyspark/ml/feature.py
Comment (felixcheung) [inline python/pyspark/ml/feature.py]: I think you need to backtick and add a tag for the attribute

File: python/pyspark/ml/feature.py
Comment (felixcheung) [inline python/pyspark/ml/feature.py]: should it say what the default value is?

Comment (actuaryzhang) [inline python/pyspark/ml/feature.py]: OK, added default value.

File: python/pyspark/ml/feature.py
Comment (actuaryzhang) [inline python/pyspark/ml/feature.py]: Added tag.

File: python/pyspark/ml/feature.py
Comment (actuaryzhang) [inline python/pyspark/ml/feature.py]: @felixcheung `stringOrderType` is not a shared trait on the Scala side, and I thought only the shared traits should be automatically generated. 
I have looked at the code for other ML transformers, and many of them hard coded, for example, [Imputer](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L933) and [OneHotEncoder](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L1520). 
Please let me know if I'm wrong, and a reference to example would be greatly appreciated. Thanks much!

**Comment by actuaryzhang** [comment]
@felixcheung Thanks so much for the review. I addressed most of the comments except auto generating code for defining `stringOrderType`. This parameter is not a shared trait on the Scala side, and I thought only the shared traits should be automatically generated. I have looked at the code for other ML transformers, and many of them hard coded, for example, [Imputer](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L933) and [OneHotEncoder](https://github.com/apache/spark/blob/master/python/pyspark/ml/feature.py#L1520).

Please let me know if I'm wrong, and a reference to example would be greatly appreciated. 
Thanks much!

File: python/pyspark/ml/feature.py
Comment (felixcheung) [inline python/pyspark/ml/feature.py]: hmm, ok, I see a few examples that they are not generated

**Comment by actuaryzhang** [comment]
@viirya @felixcheung Any additional changes needed for this one?

**Comment by felixcheung** [comment]
LGTM, ping @holdenk @jkbradley if they are interested

**Comment by viirya** [comment]
LGTM

**Comment by actuaryzhang** [comment]
@felixcheung Would you help merge this? Thanks.

**Comment by felixcheung** [comment]
I'd hold this for another 3-4 days just in case..

**Comment by MLnick** [comment]
LGTM @felixcheung

File: python/pyspark/ml/feature.py
Comment (holdenk) [inline python/pyspark/ml/feature.py]: I know were mixed on doing this, but I like including the default value in the docstring, makes the documentation closer to the Scala doc and makes it easier to read without having to refer to the ScalaDoc.

**Comment by holdenk** [comment]
One minor optional comment, but not a blocker so LGTM (although if you decide to update the docstring LGTM pending tests).

**Comment by actuaryzhang** [comment]
@holdenk  Thanks for the comment. Added default value in docstring. 
@felixcheung Please let me know if there is anything else needed for this PR. 
Thanks everyone for the review and comments!

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by felixcheung** [comment]
merged to master. thanks everyone!

**Pull request closed without merging**
